File: web/lib/freshness.ts
/**
 * Whether a stored briefing still describes the current price data.
 *
 * Mirror of `stock_analysis/synthesis/freshness.py`, which is the canonical
 * owner of this rule. It exists twice because two of the three read paths go
 * through Python (the FastAPI adapter computes and sends `briefing_stale`) but
 * the direct-Supabase path in `data.ts` derives its summary rows in the
 * browser tier and has no Python to ask. Keep the two in step: the thresholds
 * and the branch order below are the contract.
 *
 * The failure this catches: `stock-fetch` refreshes prices daily while the LLM
 * layers only rerun with the full pipeline, so a `buy` adjudicated on a
 * $592.85 close sat beside a $617.79 tape for a week and read as current.
 * Plain age does not catch it — a five-day-old briefing is fine if the stock
 * did not move, and misleading if it ran 8%.
 */

/** Beyond this, a briefing is old news even with no newer bar to compare to. */
export const STALE_AFTER_DAYS = 7;

export interface Freshness {
  stale: boolean;
  reason: string | null;
}

const FRESH: Freshness = { stale: false, reason: null };

function toUtcDay(iso: string | null | undefined): number | null {
  if (!iso) return null;
  const parsed = Date.parse(`${iso.slice(0, 10)}T00:00:00Z`);
  return Number.isNaN(parsed) ? null : parsed;
}

const DAY_MS = 86_400_000;

/**
 * @param briefingDate  When the run happened.
 * @param dataAsOf      Newest price bar the run actually read. Preferred over
 *                      `briefingDate`; absent on briefings written before the
 *                      field existed, in which case the run date stands in
 *                      (it is the same day or later, so the check stays
 *                      conservative).
 * @param latestBarDate Newest bar available now.
 */
export function assessFreshness(
  briefingDate: string | null | undefined,
  dataAsOf: string | null | undefined,
  latestBarDate: string | null | undefined,
): Freshness {
  const referenceIso = dataAsOf ?? briefingDate ?? null;
  const reference = toUtcDay(referenceIso);
  if (reference == null) {
    return { stale: true, reason: "Briefing has no date; treat as unverified." };
  }

  const latest = toUtcDay(latestBarDate);
  if (latest != null && latest > reference) {
    const lag = Math.round((latest - reference) / DAY_MS);
    return {
      stale: true,
      reason:
        `Briefing analysed price data through ${referenceIso!.slice(0, 10)}; ` +
        `bars now run to ${latestBarDate!.slice(0, 10)} (${lag}d newer). ` +
        "Its signal, conviction and levels predate the current tape — " +
        "rerun the pipeline before acting.",
    };
  }

  const now = new Date();
  const today = Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate());
  const age = Math.round((today - reference) / DAY_MS);
  if (age > STALE_AFTER_DAYS) {
    return {
      stale: true,
      reason:
        `Briefing analysed price data through ${referenceIso!.slice(0, 10)}, ` +
        `${age} days ago (limit ${STALE_AFTER_DAYS}d). ` +
        "Rerun the pipeline before acting.",
    };
  }

  return FRESH;
}

File: src/stock_analysis/synthesis/freshness.py
"""Whether a stored briefing still describes the current price data.

The pipeline's layers refresh independently: `stock-fetch` rewrites
`price_history.csv` and `technicals.json` daily, while the LLM layers
(`analyst_reports` → `debate` → `briefing`) only rerun when the full pipeline
does. Nothing previously compared the two, so a briefing adjudicated on the
2026-09-02 close sat beside technicals from 2026-09-08 and read as current: a
`buy` formed at $592.85, presented against a $617.79 tape and an RSI of 88.6
that no analyst in that run ever saw.

This module is the comparison, kept deterministic and out of the LLM layers so
a briefing cannot vouch for its own currency.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date

from stock_analysis.models.market_data import TechnicalSnapshot
from stock_analysis.models.synthesis import Briefing

# Beyond this, a briefing is old news even when no newer bar exists to compare
# it against — it matches the swing horizon the briefings are written for.
STALE_AFTER_DAYS = 7


@dataclass(frozen=True)
class Freshness:
    """Whether a briefing may be presented as current, and why not."""

    stale: bool
    reason: str | None = None

    def __bool__(self) -> bool:  # `if freshness:` reads as "is it fresh"
        return not self.stale


_FRESH = Freshness(stale=False)


def _as_date(value: str | date | None) -> date | None:
    if value is None:
        return None
    if isinstance(value, date):
        return value
    try:
        return date.fromisoformat(str(value)[:10])
    except ValueError:
        return None


def assess_freshness(
    briefing_date: str | date | None,
    data_as_of: str | date | None,
    latest_bar_date: str | date | None,
    *,
    today: date | None = None,
    stale_after_days: int = STALE_AFTER_DAYS,
) -> Freshness:
    """Compare what a briefing analysed against what the price data now shows.

    ``data_as_of`` is the newest bar the analysis actually saw and is preferred
    over ``briefing_date`` (which records when the run happened, not what it
    read). Briefings written before that field existed fall back to the run
    date, which is the same day or later, so the comparison stays conservative.
    """
    reference = _as_date(data_as_of) or _as_date(briefing_date)
    if reference is None:
        return Freshness(stale=True, reason="Briefing has no date; treat as unverified.")

    latest = _as_date(latest_bar_date)
    if latest is not None and latest > reference:
        lag = (latest - reference).days
        return Freshness(
            stale=True,
            reason=(
                f"Briefing analysed price data through {reference}; bars now run "
                f"to {latest} ({lag}d newer). Its signal, conviction and levels "
                "predate the current tape — rerun the pipeline before acting."
            ),
        )

    age = ((today or date.today()) - reference).days
    if age > stale_after_days:
        return Freshness(
            stale=True,
            reason=(
                f"Briefing analysed price data through {reference}, {age} days "
                f"ago (limit {stale_after_days}d). Rerun the pipeline before acting."
            ),
        )

    return _FRESH


def briefing_freshness(
    briefing: Briefing | None,
    technicals: TechnicalSnapshot | None = None,
    latest_bar_date: str | date | None = None,
    *,
    today: date | None = None,
) -> Freshness:
    """``assess_freshness`` for the loaded-model case."""
    if briefing is None:
        return Freshness(stale=False)
    return assess_freshness(
        briefing.date,
        briefing.data_as_of,
        latest_bar_date or (technicals.as_of_date if technicals else None),
        today=today,
    )
